Automattic/jetpack
# AI Client: update and export SuggestionsEventSource library

This PR involves the following updates:

1. It refines and exports the `SuggestionsEventSource` class from the AI Client package.
2. It introduces the `SuggestionsEventSource` into the Jetpack plugin by importing and implementing it.
3. The duplicated code is streamlined for better readability and efficiency.

From the perspective of the SuggestionsEventSource, a few non-breaking compatibility changes have been made:

- The `url` parameter is now optional, meaning passing the URL when utilizing the class is no longer mandatory.
- The class can now handle the 'fromCache' option.
- It also accepts the 'feature' option.

Take a look a the [SuggestionsEventSource doc page](https://github.com/Automattic/jetpack/blob/89d002b6aec2ebb2049e6b22ed78cb903cd52aaf/projects/js-packages/ai-client/src/suggestions-event-source/Readme.md) for more detail.

Fixes partially https://github.com/Automattic/jetpack/issues/31936
Fixes https://github.com/Automattic/jetpack/issues/31935

## Proposed changes:

* AI Client: update and export SuggestionsEventSource library

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






The best way to test this PR is simply using the AI Assistant block of the Jetpack plugin.

* Go to the block editor
* Create an AI Assistant block, or use the AI Extension tool
* Request changes.
* Confirm it works as expected.
* You can take a look at the debug logs

```
localStorage.setItem( 'debug', 'jetpack-*' )
```

The following screenshot shows how the lib defined the `url`, and also sets the `ai-assistant-experimental` feature.


<img width="850" alt="Screenshot 2023-07-18 at 15 12 51" src="https://github.com/Automattic/jetpack/assets/77539/df146db1-a4e2-4aa6-9910-beac6bd76027">


Use the AI Playground to test it:

<img width="878" alt="Screenshot 2023-07-18 at 15 13 08" src="https://github.com/Automattic/jetpack/assets/77539/f78463e4-80d2-48e0-aaa9-39e01e2cca54">

You can take a look at the request payload, too.

<img width="528" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/2cec6ab6-93f5-4f8d-952d-cb9fb47d7ba3">

**`Done` log**

<img width="835" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/625b4630-723b-4e3b-963d-f71ae72da0d1">

### Testing error response

You can hardcode the jetpack-ai-query file to force sending errors from the server side when responding to the request to be able to see how the app handles it—for instance, trying to send `422`, `429`, `500` and `503` status code.

This is a test for error code `500`:

<img width="1653" alt="Screenshot 2023-07-18 at 17 32 02" src="https://github.com/Automattic/jetpack/assets/77539/c8a8fd6e-7107-49e7-ac61-c1fb612c552f">


* Also, you can confirm the `error_unclear_prompt` (JETPACK_AI_ERROR) is properly handled.

<img width="681" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/1b806d5f-49cf-4872-bc21-bcb2a42996ae">

<img width="542" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/538b6dfc-8bd2-4ee5-9387-c7a296f48d3d">
